stdlib/public/SDK/Foundation/NSDictionary.swift

```diff
@@ -80,6 +80,23 @@ extension Dictionary : _ObjectiveCBridgeable {
       return
     }
 
+    if Key.self == String.self {
+      // String and NSString have different concepts of equality, so
+      // string-keyed NSDictionaries may generate key collisions when bridged
+      // over to Swift. See rdar://problem/35995647
+      var dict = Dictionary(minimumCapacity: d.count)
+      d.enumerateKeysAndObjects({ (anyKey: Any, anyValue: Any, _) in
```

Comment:
string keys are the most common form of NSDictionaries on iOS and macOS. This makes all bridges of NSDictionary (e.g. any API that vends them) O(N) recursively

Reply:
can we perhaps restrict the scope of the impact here to only strings that bridge across with varying composed/decomposed states?

Reply:
Hm, I don't expect a major performance difference between this and the original `_DictionaryBuilder` case below; both are eagerly bridging dictionaries in O(n) steps. (Although there is certainly room for microoptimizations; e.g., `updateValue` could be replaced with a subscript set as long as we don't check its return value.)

Should we revisit the decision about eager/lazy bridging? My impression was that eager bridging is a win here overall; but I wasn't involved in the original experiments. Cc @airspeedswift

Reply:
The mismatching hashes and String equality makes for quite a large gap between Swift and Foundation in this case. I believe we could implement a combination of lazy/deferred bridging where we pass lookups of single-line ASCII keys to `NSDictionary`, but we convert the entire dictionary on the first use of a nontrivial String key, or any time an indexing API is used. Deferred conversion requires mutating a potentially shared object, complicating implementation. I'm not entirely sure if the result would be an overall improvement for typical dictionary sizes and usage patterns, but it may be worth a try.

Reply:
However, I think this bug should be fixed sooner than that. This PR does not change the performance of dictionary bridging, but it does work around a serious (if rare) unrecoverable issue. (String-keyed `NSDictionary` instances get bridged over to `[String: Foo]` using an O(n) copy both before and after this change.)